File: coek/misc/nqueens.cpp
#include <map>
#include <vector>
#include <coek/coek.hpp>


void nqueens(coek::Model& model, int N)
{
std::vector<std::vector<coek::Variable>> x(N, std::vector<coek::Variable>(N));
for (int i=0; i<N; i++)
    for (int j=0; j<N; j++)
        x[i][j] = model.add_variable(0,1,0,true,false);

// obj
coek::Expression obj;
for (int i=0; i<N; i++)
    for (int j=0; j<N; j++)
        obj += x[i][j];
model.add_objective( obj );

// one per row
for (int i=0; i<N; i++) {
    coek::Expression c;
    for (int j=0; j<N; j++)
        c += x[i][j];
    model.add_constraint( c == 1 );
    }

// one per column
for (int j=0; j<N; j++) {
    coek::Expression c;
    for (int i=0; i<N; i++)
        c += x[i][j];
    model.add_constraint( c == 1 );
    }

// \diagonals_col
for (int i=0; i<N-1; i++) {
    coek::Expression c;
    c += x[0][i];
    for (int j=1; j<N-i; j++)
        c += x[j][i+j];
    model.add_constraint( c <= 1 );
    }
// \diagonals_row
for (int i=1; i<N-1; i++) {
    coek::Expression c;
    c += x[i][0];
    for (int j=1; j<N-i; j++)
        c += x[i+j][j];
    model.add_constraint( c <= 1 );
    }

// /diagonals_col
for (int i=1; i<N; i++) {
    coek::Expression c;
    c += x[0][i];
    for (int j=1; j<=i; j++)
        c += x[j][i-j];
    model.add_constraint( c <= 1 );
    }
// /diagonals_row
for (int i=1; i<N-1; i++) {
    coek::Expression c;
    c += x[i][N-1];
    for (int j=1; j<N-i; j++)
        c += x[i+j][N-1-j];
    model.add_constraint( c <= 1 );
    }
}


void nqueens_400(coek::Model& model)
{ nqueens(model, 400); }

void nqueens_600(coek::Model& model)
{ nqueens(model, 600); }

void nqueens_800(coek::Model& model)
{ nqueens(model, 800); }

void nqueens_1000(coek::Model& model)
{ nqueens(model, 1000); }


std::map <std::string, void (*)(coek::Model&)> nqueens_tests {
        {"nqueens_400", nqueens_400},
        {"nqueens_600", nqueens_600},
        {"nqueens_800", nqueens_800},
        {"nqueens_1000", nqueens_1000} };

void nqueens_instance(coek::Model& model, const std::string& name)
{ return nqueens_tests[name](model); }

File: poek/models/nqueens.py
import sys
import itertools
import poek as pk

quicksum = pk.quicksum


def nqueens(N):
    model = pk.model()

    x = model.add_variable((N,N), lb=0.0, ub=1.0, binary=True)

    # obj
    model.add_objective( quicksum(x[i,j] for i in range(N) for j in range(N)) )

    # one per row
    for i in range(N):
        model.add_constraint( quicksum(x[i,j] for j in range(N)) == 1 )

    # one per column
    for j in range(N):
        model.add_constraint( quicksum(x[i,j] for i in range(N)) == 1 )

    # \diagonals_col
    for i in range(N-1):
        model.add_constraint( x[0, i] + quicksum(x[j, i+j] for j in range(1, N-i)) <= 1 )
    # \diagonals_row
    for i in range(1, N-1):
        model.add_constraint( x[i, 0] + quicksum(x[i+j, j] for j in range(1, N-i)) <= 1 )

    # /diagonals_col
    for i in range(1,N):
        model.add_constraint( x[0, i] + quicksum(x[j, i-j] for j in range(1, i+1)) <= 1 )
    # /diagonals_row
    for i in range(1,N-1):
        model.add_constraint( x[i, N-1] + quicksum(x[i+j, N-1-j] for j in range(1, N-i)) <= 1)

    return model

